REGRESSION (262875@main / iOS 17): fill: 'both' not respected with animation

graouts opened the pull request:
#### 6f928cc543c3d8287b87fb3c38cf8ca4f5bc2829
<pre>
REGRESSION (262875@main / iOS 17): fill: &apos;both&apos; not respected with animation
<a href="https://bugs.webkit.org/show_bug.cgi?id=257861">https://bugs.webkit.org/show_bug.cgi?id=257861</a>
rdar://111407099

Reviewed by Simon Fraser.

While we fixed a similar issue with 265498@main, the fix was not complete. Indeed, to correctly
deal with accelerated replaced effects in an effect stack, we must ensure two things:

1. in `KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties()`, where we ensure
   the accelerated effect is present on the backing GraphicsLayerCA in the right sort order, we
   not only need to add an `UpdateProperties` accelerated action but also ensure that we mark it
   as running by setting the `m_lastRecordedAcceleratedAction` member.

2. in `KeyframeEffectStack::applyPendingAcceleratedActions()`, where we update all effects in the
   stack to ensure any pending accelerated actions as applied, we must *not* force the running of
   replaced effects if all the accelerated effects in the stack are *not* actively running. In that
   case we must only call `KeyframeEffect::applyPendingAcceleratedActions` to ensure any previously-running
   accelerated effect has a chance to be stopped.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-opacity-replaced-effect-in-shadow-root-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-opacity-replaced-effect-in-shadow-root-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-opacity-replaced-effect-in-shadow-root.html: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyPendingAcceleratedActions const):

Canonical link: <a href="https://commits.webkit.org/265909@main">https://commits.webkit.org/265909@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/446e2d0ba7ea2ec1e812555b0c69295c6d70452d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12580 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12883 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13980 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/11808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12265 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12596 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/14481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12399 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/13195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/10362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14398 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/10480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/11103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/11561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/11265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/14448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/11760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/9703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10976 "Failed to checkout and rebase branch from PR 15693") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/10978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/15306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1363 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/KeyframeEffectStack.cpp:
```diff
@@ -277,8 +277,16 @@ void KeyframeEffectStack::cascadeDidOverrideProperties(const HashSet<AnimatableP
 {
-    for (auto& effect : m_effects)
-        effect->applyPendingAcceleratedActionsOrUpdateTimingProperties();
+    auto hasActiveAcceleratedEffect = m_effects.containsIf([](const auto& effect) {
```
I don't find 'auto' useful for bool.

Shame we have to do two passes over m_effects in this function.

graouts replied:
> I don't find 'auto' useful for bool.

I'll use `bool` when landing.

> Shame we have to do two passes over m_effects in this function.

True. That said, it's unlikely that `m_effects` is ever containing a large number of items. These are only the [_relevant_](https://drafts.csswg.org/web-animations-1/#relevant-animations-section)  effects for a given element. Also, we don't run this function on each frame, only when we expect a change in running state for an accelerated effect in the stack, which "naturally" would only happen as effects become or stop being relevant.

graouts committed:
REGRESSION (262875@main / iOS 17): fill: 'both' not respected with animation
https://bugs.webkit.org/show_bug.cgi?id=257861
rdar://111407099

Reviewed by Simon Fraser.

While we fixed a similar issue with 265498@main, the fix was not complete. Indeed, to correctly
deal with accelerated replaced effects in an effect stack, we must ensure two things:

1. in `KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties()`, where we ensure
   the accelerated effect is present on the backing GraphicsLayerCA in the right sort order, we
   not only need to add an `UpdateProperties` accelerated action but also ensure that we mark it
   as running by setting the `m_lastRecordedAcceleratedAction` member.

2. in `KeyframeEffectStack::applyPendingAcceleratedActions()`, where we update all effects in the
   stack to ensure any pending accelerated actions as applied, we must *not* force the running of
   replaced effects if all the accelerated effects in the stack are *not* actively running. In that
   case we must only call `KeyframeEffect::applyPendingAcceleratedActions` to ensure any previously-running
   accelerated effect has a chance to be stopped.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-opacity-replaced-effect-in-shadow-root-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-opacity-replaced-effect-in-shadow-root-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-opacity-replaced-effect-in-shadow-root.html: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyPendingAcceleratedActions const):

Canonical link: https://commits.webkit.org/265909@main
